File: src/main/java/com/timeback/ui/feature/timeline/TimelineViewModel.java
package com.timeback.ui.feature.timeline;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.timeback.ui.domain.gateway.FeatureGateway;
import com.timeback.ui.domain.model.*;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

/**
 * UI-03 타임라인 ViewModel.
 * 날짜별 앱 사용 타임라인을 로드하고, 활동 생성/수정/맥락 확인을 처리한다.
 */
@HiltViewModel
public class TimelineViewModel extends ViewModel {

    private final FeatureGateway gateway;
    private final MutableLiveData<ScreenState> _state = new MutableLiveData<>(new ScreenState.Initial());
    public final LiveData<ScreenState> state = _state;

    private final MutableLiveData<ActionResult> _actionResult = new MutableLiveData<>();
    public final LiveData<ActionResult> actionResult = _actionResult;

    private long currentDate = System.currentTimeMillis();

    @Inject
    public TimelineViewModel(FeatureGateway gateway) {
        this.gateway = gateway;
        loadTimeline(currentDate);
    }

    public void loadTimeline(long date) {
        this.currentDate = date;
        _state.setValue(new ScreenState.Loading());
        gateway.readTimeline(date).thenAccept(_state::postValue);
    }

    public void refresh() {
        gateway.refreshTimeline(currentDate).thenAccept(_state::postValue);
    }

    public void createActivity(ActivityType type, String name, TimeRange timeRange) {
        gateway.createActivity(type, name, timeRange).thenAccept(_actionResult::postValue);
    }

    public void updateActivity(Identifier activityId, String name, TimeRange timeRange) {
        gateway.updateActivity(activityId, name, timeRange).thenAccept(_actionResult::postValue);
    }

    public void confirmContext(Identifier contextId) {
        gateway.confirmContext(contextId).thenAccept(_actionResult::postValue);
    }

    public void updateContext(Identifier contextId, ContextClassification classification) {
        gateway.updateContext(contextId, classification).thenAccept(_actionResult::postValue);
    }
}

File: src/main/java/com/timeback/ui/feature/timeline/TimelineViewModel.kt
package com.timeback.ui.feature.timeline

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.timeback.ui.domain.gateway.FeatureGateway
import com.timeback.ui.domain.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI-03 Timeline ViewModel
 * CT-04 §6: ReadTimeline, RefreshTimeline, CreateActivity, UpdateActivity, ConfirmContext, UpdateContext
 */
@HiltViewModel
class TimelineViewModel @Inject constructor(
    private val gateway: FeatureGateway
) : ViewModel() {

    private val _state = MutableStateFlow<ScreenState>(ScreenState.Initial)
    val state: StateFlow<ScreenState> = _state.asStateFlow()

    private val _actionResult = MutableStateFlow<ActionResult?>(null)
    val actionResult: StateFlow<ActionResult?> = _actionResult.asStateFlow()

    private var currentDate: Long = System.currentTimeMillis()

    fun loadTimeline(date: Long) {
        currentDate = date
        viewModelScope.launch {
            _state.value = ScreenState.Loading
            _state.value = gateway.readTimeline(date)
        }
    }

    fun refreshTimeline() {
        viewModelScope.launch {
            val current = _state.value
            if (current is ScreenState.Content<*>) {
                _state.value = ScreenState.Refreshing(current.data)
            }
            _state.value = gateway.refreshTimeline(currentDate)
        }
    }

    fun createActivity(type: ActivityType, name: String, timeRange: TimeRange) {
        viewModelScope.launch {
            val result = gateway.createActivity(type, name, timeRange)
            _actionResult.value = result
            if (result is ActionResult.Success) {
                _state.value = gateway.readTimeline(currentDate)
            }
        }
    }

    fun updateActivity(activityId: Identifier, name: String?, timeRange: TimeRange?) {
        viewModelScope.launch {
            val result = gateway.updateActivity(activityId, name, timeRange)
            _actionResult.value = result
            if (result is ActionResult.Success) {
                _state.value = gateway.readTimeline(currentDate)
            }
        }
    }

    fun confirmContext(contextId: Identifier) {
        viewModelScope.launch {
            val result = gateway.confirmContext(contextId)
            _actionResult.value = result
            if (result is ActionResult.Success) {
                _state.value = gateway.readTimeline(currentDate)
            }
        }
    }

    fun updateContext(contextId: Identifier, classification: ContextClassification) {
        viewModelScope.launch {
            val result = gateway.updateContext(contextId, classification)
            _actionResult.value = result
            if (result is ActionResult.Success) {
                _state.value = gateway.readTimeline(currentDate)
            }
        }
    }

    fun clearActionResult() {
        _actionResult.value = null
    }
}
